// @splitsoftware/splitio/lib/storage/EventsCache/InRedis.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _logger = _interopRequireDefault(require("../../utils/logger"));

/**
Copyright 2016 Split Software

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
var log = (0, _logger.default)('splitio-storage:redis');

var EventsCache = /*#__PURE__*/function () {
  function EventsCache(keys, redis, meta) {
    this.keys = keys;
    this.redis = redis;
    this.meta = meta;
    this.eventsKey = keys.buildEventsKey();
  }
  /**
   * Add a new event object into the queue.
   */


  var _proto = EventsCache.prototype;

  _proto.track = function track(eventData) {
    return this.redis.rpush(this.eventsKey, this._toJSON(eventData)) // We use boolean values to signal successful queueing
    .then(function () {
      return true;
    }).catch(function (err) {
      log.error("Error adding event to queue: " + err + ".");
      return false;
    });
  }
  /**
   * Generates the JSON as we'll store it on Redis.
   */
  ;

  _proto._toJSON = function _toJSON(eventData) {
    return JSON.stringify({
      m: this.meta,
      e: eventData
    });
  }
  /**
   * We are returning true because the go syncronizer push the impressions from redis.
   */
  ;

  _proto.isEmpty = function isEmpty() {
    return true;
  };

  return EventsCache;
}();

var _default = EventsCache;
exports.default = _default;
