// @splitsoftware/splitio/lib/storage/CountCache/InMemory.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

/**
Copyright 2016 Split Software

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
var CountCacheInMemory = /*#__PURE__*/function () {
  function CountCacheInMemory() {
    this.clear();
  }

  var _proto = CountCacheInMemory.prototype;

  _proto.clear = function clear() {
    this.counters = {};
    return this;
  };

  _proto.state = function state() {
    return this.counters;
  };

  _proto.track = function track(metricName) {
    if (this.counters[metricName] === undefined) this.counters[metricName] = 1;else this.counters[metricName]++;
    return this.counters[metricName];
  };

  _proto.isEmpty = function isEmpty() {
    return Object.keys(this.counters).length === 0;
  };

  _proto.toJSON = function toJSON() {
    return this.counters;
  };

  return CountCacheInMemory;
}();

var _default = CountCacheInMemory;
exports.default = _default;
